### ISSUE TITLE
ArrayStride decoration not transferred to function parameters

### ISSUE THREAD

**Issue by gnl21** [google/clspv#637]
For the test kernel:
```
static int test_func(const int idx, global const unsigned char (*const d)[6])
{
  return d[idx][0];
}

kernel void k1(write_only image2d_t out, global float *padding, global const unsigned char (*const data)[6])
{
  int d = test_func(0, data);
  write_imagei(out, (int2)(0, 0), d);
}

kernel void k2(write_only image2d_t out, global const unsigned char (*const data)[6])
{
  int d = test_func(0, data);
  write_imagei(out, (int2)(0, 0), d);
}
```
clspv appears to generate invalid SPIR-V for `test_func`. The function parameter `d` is not decorated with `ArrayStride` so the consumer can't generate the correct memory accesses.

If I simplify the kernel (remove either `k1` or `k2`, or remove the `padding` argument) then the SPIR-V is transformed so that `data` is not passed as an argument at all, working around the problem.

**Comment by alan-baker** [comment]
I'm not there's actually a problem here. The only rules about ArrayStride I can see in the spec are that arrays in transparent storage classes require the decoration and any pointer used as the base of an `OpPtrAccessChain` requires the decoration. The code snippet you provided doesn't utilize OpPtrAccessChain. It only has a regular access chain before the function call and another in the sub-function.

When I compile the following (`clspv test.cl -o test.spv -no-inline-single -no-dra`), I do see the pointer parameter decorated with ArrayStride due to the OpPtrAccessChain in `bar`:
```
void bar(global int* data, int n) {
  data[1] = n;
}

kernel void foo(global int* data, int n) {
  bar(data, n);
}
```

Which rule in the spec do you think the sample code violates?

**Comment by gnl21** [comment]
> Which rule in the spec do you think the sample code violates?

I considered filing this in the spec repo, but looking through the issues there I convinced myself that this is supposed to be required. Without these decorations we have to compile separate version of these functions at call sites once we know the layout information of the buffer that the pointer is pointing into. I'll open a spec issue to clarify it.

**Comment by gnl21** [comment]
OK, I've looked over this again and it looks like I might have simplified the problem away, then got confused. Let me try again. Compiling the test kernel:
```
void bar(const int row, global unsigned char (*data)[6])
{
  data[row][0] = 0;
}

kernel void foo(global unsigned char (*data)[6])
{
  bar(0, data);
}
```
with `clspv -no-dra test.cl` gives me SPIR-V that uses `OpPtrAccessChain` on a function parameter but the type of the parameter does not have an `ArrayStride` decoration. (The pointer is to the first element of the runtime array `data`, and the runtime array type is decorated `ArrayStride 6`, but the pointer-to-first-element type is not).

**Comment by gnl21** [comment]
Also, I think that the rule:
> Pointer types are also allowed to have multiple \<id\>s for the same opcode and operands, to allow for differing decorations (e.g., Volatile) or different decoration values (e.g., different Array Stride values for the ArrayStride). When new pointers are formed, their types must be decorated as needed, so the consumer knows how to generate an access through the pointer.

is intended to require that variable pointers be decorated with ArrayStride even if they are not being used for `OpPtrAccessChain`.

**Comment by alan-baker** [comment]
Thanks for the updated testcase. I agree clspv seems to be generating the ArrayStride decorations incorrectly. For some reason it is decorating the result of the OpPtrAccessChain instead of the base pointer. That would address this particular case, but I want to understand your driver's requirements for array strides more generally. If the original testcase you filed doesn't require an array stride on the parameter then other than as the base of pointer access chains, where do you require an array stride? My understanding is **only** the base requires an array stride to generate the correct address when considering the *Element* operand.

**Comment by gnl21** [comment]
I think that my confusion with the original testcase may have made it seem like there's a problem where none may actually exist. At least, I hope so. I don't currently have any failing test case other than this new one. (I'm sure the old one was failing for me at some point, but it is not with the latest clspv code).

Our implementation requires an `ArrayStride` whenever it is indexing a pointer into an array, whether as the base of an `OpPtrAccessChain` or from any of the other indices in any access chain. That is, variable pointers do not carry any stride information with them, it is taken from the decorations. Most of the time this information is available because the type of the pointer mostly inherits the type hierarchy of the original object. It sounds plausible that the only time the pointer's type does not refer back to the decoration in the original object's type is when that pointer is being used as part of an `OpPtrAccessChain`, but I'll need to think about it a little bit more. Does that make sense?

**Comment by alan-baker** [comment]
I think that is true. The array stride would normally be decorated on the array type itself (in a transparent storage class), but OpPtrAccessChain allows for treating an element of the array as an array pointer (effectively) which is where the extra requirement arises that the base pointer also requires the array stride decoration.

I just wanted to clarify this because I believe the fix is small in clspv to address this bug, but if clspv is missing other array stride decorations, I need to put some more thought into it.
